File: src/main/scala/L1Compiler/L1AST.scala
package L1Compiler

object L1AST {

  object L1{ def apply(main: Func): L1 = L1(main, Nil) }
  case class L1(main: Func, funs:List[Func]){
    def toCode = "(" + main.toCode(false) + "\n" + funs.map(_.toCode(true)).mkString("\n") + ")"
  }
  case class Func(name: LabelDeclaration, body: List[Instruction]){
    def toCode(printLabel:Boolean=true) = {
      if(printLabel) "(" + name.l.toCode + "\n" + body.map(_.toCode).mkString("\n") + ")"
      else "(" + body.map(_.toCode).mkString("\n") + ")"
    }
  }

  sealed trait AssignmentRHS { def toCode:String }

  case class Allocate(n:S, init:S) extends AssignmentRHS {
    def toCode: String = "(allocate " + n.toCode + " " + init.toCode + ")"
  }
  case class Assignment(r:Register, rhs:AssignmentRHS) extends Instruction {
    def toCode: String = "(" + r.toCode + " <- " + rhs.toCode + ")"
  }
  case class Increment(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + r.toCode + " += " + s.toCode + ")"
  }
  case class Decrement(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + r.toCode + " -= " + s.toCode + ")"
  }
  case class Multiply(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + r.toCode + " *= " + s.toCode + ")"
  }
  case class LeftShift(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + r.toCode + " <<= " + s.toCode + ")"
  }
  case class RightShift(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + s.toCode + " >>= " + r.toCode + ")"
  }
  case class BitwiseAnd(r:Register, s:S) extends Instruction {
    def toCode: String = "(" + r.toCode + " &= " + s.toCode + ")"
  }
  case class MemLoc(basePointer:Register, offset: Num) {
    def toCode: String = "(mem " + basePointer.toCode + " " + offset.toCode + ")"
  }
  case class MemRead(loc: MemLoc) extends AssignmentRHS {
    def toCode: String = loc.toCode
  }
  case class MemWrite(loc: MemLoc, s:S) extends Instruction {
    def toCode: String = "(" + loc.toCode + " <- " + s.toCode + ")"
  }
  case class Print(s:S) extends AssignmentRHS {
    def toCode: String = "(print " + s.toCode + ")"
  }
  // TODO: check if interpreter allows (goto num) and (goto register)
  case class Goto(s:S) extends Instruction {
    def toCode: String = "(goto " + s.toCode + ")"
  }
  case class CJump(comp:Comp, l1: Label, l2: Label) extends Instruction {
    def toCode: String = "(cjump " + comp.toCode + " " + l1.toCode  + " " + l2.toCode + ")"
  }
  case class Call(s:S) extends Instruction {
    def toCode: String = "(call " + s.toCode + ")"
  }
  case class TailCall(s:S) extends Instruction {
    def toCode: String = "(call " + s.toCode + ")"
  }
  case object Return extends Instruction {
    def toCode: String = "(return)"
  }
  case class ArrayError(s1:S, s2:S) extends AssignmentRHS {
    def toCode: String = "(array-error " + s1.toCode + " " + s2.toCode + ")"
  }

  sealed trait Instruction{
    def toCode: String
  }
  sealed trait S extends AssignmentRHS
  case class Num(n: Int) extends S {
    def toCode: String = n.toString
  }
  case class Label(name: String) extends S {
    override def toString = "Label(\"" + name + "\")"
    def toCode: String = ":" + name
  }
  case class LabelDeclaration(l: Label) extends Instruction {
    def toCode: String = l.toCode
  }

  sealed trait Register extends S {
    val name: String
    override def toString = name
    def toCode: String = name
  }
  object XRegister {
    def apply(s: Symbol): Option[XRegister] = s match {
      case 'esi => Some(esi)
      case 'edi => Some(edi)
      case 'ebp => Some(ebp)
      case 'esp => Some(esp)
      case _ => None
    }
  }
  sealed abstract case class XRegister(name: String) extends Register
  object esi extends XRegister("esi")
  object edi extends XRegister("edi")
  object ebp extends XRegister("ebp")
  object esp extends XRegister("esp")
  object CXRegister {
    def apply(s: Symbol): Option[CXRegister] = s match {
      case 'eax => Some(eax)
      case 'ecx => Some(ecx)
      case 'edx => Some(edx)
      case 'ebx => Some(ebx)
      case _ => None
    }
  }
  sealed abstract case class CXRegister(name: String) extends Register{
    def low8 = "%" + name(1) + "l"
  }
  object eax extends CXRegister("eax")
  object ecx extends CXRegister("ecx")
  object edx extends CXRegister("edx")
  object ebx extends CXRegister("ebx")

  // TODO: this is DEFINITELY not an instruction
  // its only part of two different instructions
  case class Comp(s1:S, op: CompOp, s2:S) extends AssignmentRHS {
    def toCode: String = s1.toCode + " " + op.op + " " + s2.toCode
  }
  sealed abstract case class CompOp(op: String){
    def apply(x:Int, y:Int): Boolean
  }
  object LessThan extends CompOp("<"){
    override def toString = "LessThan"
    def apply(x:Int, y:Int) = x < y
  }
  object LessThanOrEqualTo extends CompOp("<="){
    override def toString = "LessThanOrEqualTo"
    def apply(x:Int, y:Int) = x <= y
  }
  object EqualTo extends CompOp("="){
    override def toString = "EqualTo"
    def apply(x:Int, y:Int) = x == y
  }
}

File: src/main/haskell/L1AST.hs
module L1AST where

import TestHelpers
import Test.HUnit
import Control.Monad

data XRegister  = Esi | Edi | Ebp | Esp
data CXRegister = Eax | Ebx | Ecx | Edx
data Register   = CXR CXRegister | XR XRegister

type Label = String
data MemLoc x = MemLoc x Int
data CompOp = LT | LTEQ | EQ
data Comp s = Comp s CompOp s

data AssignRHS x s = 
  CompRHS (Comp s)   | 
  Allocate s s       |
  Print s            |
  ArrayError s s     |
  MemRead (MemLoc x) |
  SRHS s

data Instruction x s = 
  Assign x (AssignRHS x s)   |
  Increment x s              |
  Decrement x s              |
  Multiply x s               |
  LeftShift x s              |
  RightShift x s             |
  BitwiseAnd x s             |
  MemWrite (MemLoc x) s      |
  Goto Label                 |
  CJump (Comp s) Label Label |
  LabelDeclaration Label     |
  Call s                     |
  TailCall s                 |
  Return 

type L1X = Register
data L1S = NumberL1S Int | LabelL1S Label | RegL1S Register
type L1Instruction = Instruction L1X L1S
data L1Func = L1Func {body :: [L1Instruction]}
data L1 = L1 L1Func [L1Func]

instance Show L1 where
  show (L1 main fs) = "(\n" ++ (show main) ++ "\n" ++ (fs >>= show) ++ "\n)"

instance Show L1Func where
  show (L1Func is) = "(" ++ (is >>= (\i -> ((show i) ++ "\n\t"))) ++ ")"

instance (Show x, Show s) => Show (Instruction x s) where
  show (Assign x rhs)       = "(" ++ (show x) ++ " <- "   ++ (show rhs) ++ ")"
  show (Increment x s)      = "(" ++ (show x) ++ " += "   ++ (show s) ++ ")"
  show (Decrement x s)      = "(" ++ (show x) ++ " -= "   ++ (show s) ++ ")"
  show (Multiply x s)       = "(" ++ (show x) ++ " *= "   ++ (show s) ++ ")"
  show (LeftShift x s)      = "(" ++ (show x) ++ " <<= "  ++ (show s) ++ ")"
  show (RightShift x s)     = "(" ++ (show x) ++ " >>= "  ++ (show s) ++ ")"
  show (BitwiseAnd x s)     = "(" ++ (show x) ++ " &= "   ++ (show s) ++ ")"
  show (MemWrite loc s)     = "(" ++ (show loc) ++ " <- " ++ (show s) ++ ")"
  show (Goto l)             = "(goto" ++ (show l) ++ ")"
  show (CJump cmp l1 l2)    = join ["(", (show cmp), " ", (show l1), " ", (show l2), ")"]
  show (LabelDeclaration l) = show l
  show (Call s)             = "(call" ++ (show s) ++ ")"
  show (TailCall s)         = "(tail-call" ++ (show s) ++ ")"
  show Return               = "(return)"

instance Show L1S where
  show (NumberL1S n) = show n
  show (LabelL1S l) = show l
  show (RegL1S r) = show r

instance Show XRegister where
  show Esi = "esi"
  show Edi = "edi"
  show Ebp = "ebp"
  show Esp = "esp"
instance Show CXRegister where
  show Eax = "eax"
  show Ebx = "ebx"
  show Ecx = "ecx"
  show Edx = "edx"

instance Show Register where
  show (CXR cxr) = show cxr
  show (XR xr) = show xr

instance (Show x, Show s) => Show (AssignRHS x s) where
  show (CompRHS c)        = show c
  show (Allocate s1 s2)   = "(allocate " ++ (show s1) ++ " " ++ (show s2) ++ ")"
  show (Print s)          = "(print " ++ (show s) ++ ")"
  show (ArrayError s1 s2) = "(array-error " ++ (show s1) ++ " " ++ (show s2) ++ ")"
  show (SRHS s)           = show s
  show (MemRead loc)      = show loc

instance (Show x) => Show (MemLoc x) where
  show (MemLoc x n) = "(mem " ++ (show x) ++ " " ++ (show n) ++ ")"

instance (Show s) => Show (Comp s) where
  show (Comp s1 op s2) = (show s1) ++ " " ++ (show op) ++ " " ++ (show s2)

xRegisters      = [Esi, Edi, Ebp, Esp]
cxRegisters     = [Eax, Ebx, Ecx, Edx]
xRegisterNames  = map show xRegisters
cxRegisterNames = map show cxRegisters

xRegisterFromName :: String -> Maybe XRegister
xRegisterFromName "esi" = Just Esi
xRegisterFromName "edi" = Just Edi
xRegisterFromName "ebp" = Just Ebp
xRegisterFromName "esp" = Just Esp
xRegisterFromName _ = Nothing

cxRegisterFromName :: String -> Maybe CXRegister
cxRegisterFromName "eax" = Just Eax
cxRegisterFromName "ebx" = Just Ebx
cxRegisterFromName "ecx" = Just Ecx
cxRegisterFromName "edx" = Just Edx
cxRegisterFromName _ = Nothing

registerFromName :: String -> Maybe Register
registerFromName name = case (xRegisterFromName name) of
  Just xr -> Just (XR xr)
  Nothing -> fmap CXR (cxRegisterFromName name)

instance Show CompOp where
  show L1AST.LT = "<"
  show LTEQ     = "<="
  show L1AST.EQ = "="

compOpFromSym :: String -> CompOp
compOpFromSym "<"  = L1AST.LT
compOpFromSym "<=" = LTEQ
compOpFromSym "="  = L1AST.EQ
compOrFromSym _ = error "not a comparison operator"

runOp L1AST.LT   n1 n2 = n1 <  n2
runOp L1AST.LTEQ n1 n2 = n1 <= n2
runOp L1AST.EQ   n1 n2 = n1 == n2
